kernel/src/grant.rs

```diff
@@ -129,9 +130,207 @@ use core::slice;
+    allow_ro_num: *mut usize,
+    allow_ro_array: *mut SavedAllowRo,
+    allow_rw_num: *mut usize,
+    allow_rw_array: *mut SavedAllowRw,
+}
+
+/// Represents the size of the upcall array in the kernel owned section of the grant
+#[derive(Copy, Clone)]
+struct UpcallItems(usize);
+/// Represents the size of the allow read-only array in the kernel owned section of the grant
+#[derive(Copy, Clone)]
+struct AllowRoItems(usize);
+/// Represents the size of the allow read-write array in the kernel owned section of the grant
+#[derive(Copy, Clone)]
+struct AllowRwItems(usize);
+/// Represents the size data T within the grant
+#[derive(Copy, Clone)]
+struct GrantDataSize(usize);
+/// Represents the alignment of data T within the grant
+#[derive(Copy, Clone)]
+struct GrantDataAlign(usize);
+
+impl KernelManagedLayout {
+    /// Reads the specified pointer as the base of the kernel owned grant region.
+    ///
+    /// # Safety
+    ///
+    /// The incoming base pointer must be well aligned and already contain initialized data
+    /// in the expected form.
+    unsafe fn read_from_base(base_ptr: *mut u8) -> Self {
+        let upcalls_num = base_ptr as *mut usize;
+        let upcalls_array = upcalls_num.add(1) as *mut SavedUpcall;
+
+        let allow_ro_num = upcalls_array.add(upcalls_num.read()) as *mut usize;
+        let allow_ro_array = allow_ro_num.add(1) as *mut SavedAllowRo;
+
+        let allow_rw_num = allow_ro_array.add(allow_ro_num.read()) as *mut usize;
+        let allow_rw_array = allow_rw_num.add(1) as *mut SavedAllowRw;
+
+        Self {
+            upcalls_num,
+            upcalls_array,
+            allow_ro_num,
+            allow_ro_array,
+            allow_rw_num,
+            allow_rw_array,
+        }
+    }
+
+    /// Creates a layout from the specified pointer and lengths of arrays
+    ///
+    /// # Safety
+    ///
+    /// The incoming base pointer must be well aligned but does not need to point
+    /// to initialized data
+    unsafe fn from_counts(
+        base_ptr: *mut u8,
+        upcalls_num_val: UpcallItems,
+        allow_ro_num_val: AllowRoItems,
+    ) -> Self {
+        let upcalls_num = base_ptr as *mut usize;
+        let upcalls_array = upcalls_num.add(1) as *mut SavedUpcall;
+
+        let allow_ro_num = upcalls_array.add(upcalls_num_val.0) as *mut usize;
+        let allow_ro_array = allow_ro_num.add(1) as *mut SavedAllowRo;
+
+        let allow_rw_num = allow_ro_array.add(allow_ro_num_val.0) as *mut usize;
+        let allow_rw_array = allow_rw_num.add(1) as *mut SavedAllowRw;
+
+        Self {
+            upcalls_num,
+            upcalls_array,
+            allow_ro_num,
+            allow_ro_array,
+            allow_rw_num,
+            allow_rw_array,
+        }
+    }
+
+    /// Returns the entire grant size including the kernel own memory, padding, and data for T.
+    /// Requires that grant_t_align be a power of 2, which is guaranteed from align_of rust calls
+    fn grant_size(
```

Comment:
I don't think that's what `const` does -- `const` just indicates the function can be used in a `const` context such as const generics and array sizes. E.g. you can make a reference to a `const` function, which requires the compiler to emit it as its own function.

Reply:
Oh, you're right. I thought that it would also force Rust to evaluate the function at compile time if the arguments are known at compile time but that guarantee does not exist.

Marking resolved.